### PR TITLE
Refactor parser::parse_arrow_function_body_impl issue #638

### PR DESCRIPTION
**Pull request by Rebraws**
Refactored the parse_arrow_function_body_impl  member function by merging both expression::arrow_function_with_expression and expression::arrow_function_with_statement  classes into one, also removed the temporary expression from the function parse_arrow_function_body_impl to avoid double parsing of parameters.

I think the last item mentioned in issue #638 it's already done in this commit https://github.com/quick-lint/quick-lint-js/commit/35a362bce7e9b17252da3a0c1d11237ca9268595

### CODE REVIEW

**Comment by Rebraws** [comment]
I have read and hereby agree to quick-lint-js' Contributor License Agreement (CLA-v1.md).

**Review by strager — CHANGES_REQUESTED** [review]
Must fix: See build errors reported by CI.

Your changes look good. Here's what I would like to do:

1. After fixing CI errors, merge your first commit (0034dab1bf862300d3f237252334faac86c2ab3f).
2. Merge your other commits squashed together (optionally with modifications if needed).

File: src/quick-lint-js/parse.h [hunk 1]
```diff
@@ -367,22 +367,17 @@ class parser {
...
   template <QLJS_PARSE_VISITOR Visitor>
-  expression *parse_arrow_function_body(Visitor &, function_attributes,
-                                        const char8 *parameter_list_begin,
-                                        bool allow_in_operator);
-  template <QLJS_PARSE_VISITOR Visitor>
   expression *parse_arrow_function_body(
+      Visitor &, function_attributes, const char8 *parameter_list_begin,
+      bool allow_in_operator,
+      expression_arena::array_ptr<expression *> &&parameters =
+          expression_arena::array_ptr<expression *>());
```
Comment (strager) [inline src/quick-lint-js/parse.h:374]: Nit: Avoid default parameters. Have the caller pass an `array_ptr` explicitly.

File: src/quick-lint-js/parse-expression-inl.h [hunk 2]
```diff
@@ -1779,15 +1779,6 @@ expression* parser::parse_index_expression_remainder(Visitor& v,
...
-expression* parser::parse_arrow_function_body(Visitor& v,
-                                              function_attributes attributes,
-                                              const char8* parameter_list_begin,
-                                              bool allow_in_operator) {
-  return this->parse_arrow_function_body_impl(
-      v, attributes, parameter_list_begin, allow_in_operator);
-}
-
 template <QLJS_PARSE_VISITOR Visitor>
 expression* parser::parse_arrow_function_body(
```
Comment (strager) [inline src/quick-lint-js/parse-expression-inl.h:1781]: This function is now useless. Inline `parse_arrow_function_body_impl` into this function.

**Review by strager — APPROVED** [review]

File: src/quick-lint-js/parse-expression-inl.h (hunk 2 above)
Comment (Rebraws) [reply src/quick-lint-js/parse-expression-inl.h:1783]: Solved this in the last uploaded commit https://github.com/quick-lint/quick-lint-js/pull/680/commits/1c488ed74bc8529521ee132c59122d98152563ef and also removed the default parameters in the parse_arrow_function_body function declaration

**Pull request merged**